Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp

```diff
@@ -0,0 +1,168 @@
+
+    m_workQueue->dispatch([protectedThis = Ref { *this }, protectedWorkQueue = Ref { *m_workQueue }, protectedFrameProvider = Ref { m_frameProvider }] () mutable {
+        Request request;
+        while (protectedThis->requestQueue().dequeue(request)) {
+            TraceScope tracingScope(AsyncImageDecodeStart, AsyncImageDecodeEnd);
+
+            auto& decoder = protectedFrameProvider->decoder();
+            auto minimumDecodingDuration = protectedThis->minimumDecodingDurationForTesting();
+
+            MonotonicTime startingTime;
+            if (minimumDecodingDuration > 0_s)
+                startingTime = MonotonicTime::now();
+
+            auto platformImage = decoder.createFrameImageAtIndex(request.index, request.subsamplingLevel, request.options);
+            auto nativeImage = NativeImage::create(WTFMove(platformImage));
+
+            // Pretend as if decoding the frame took minimumDecodingDuration.
+            if (minimumDecodingDuration > 0_s) {
+                auto actualDecodingDuration = MonotonicTime::now() - startingTime;
+                if (minimumDecodingDuration > actualDecodingDuration)
+                    sleep(minimumDecodingDuration - actualDecodingDuration);
+            }
+
+            // Even if we fail to decode the frame, it is important to sync the main thread with this result.
+            callOnMainThread([protectedThis, protectedWorkQueue, protectedFrameProvider, request, nativeImage = WTFMove(nativeImage)] () mutable {
+                // The WorkQueue may have been recreated before the frame was decoded.
+                if (protectedWorkQueue.ptr() != protectedThis->m_workQueue || protectedFrameProvider.ptr() != &protectedThis->m_frameProvider) {
+                    LOG(Images, "ImageFrameProvider::%s - %p - url: %s. WorkQueue was recreated at index = %d.", __FUNCTION__, protectedThis.ptr(), protectedFrameProvider->sourceUTF8(), request.index);
+                    return;
+                }
+
+                // The DecodeQueue may have been cleared before the frame was decoded.
+                if (protectedThis->decodeQueue().isEmpty() || protectedThis->decodeQueue().first() != request) {
+                    LOG(Images, "ImageFrameProvider::%s - %p - url: %s. DecodeQueue was cleared at index = %d.", __FUNCTION__, protectedThis.ptr(), protectedFrameProvider->sourceUTF8(), request.index);
+                    return;
+                }
+
+                protectedThis->decodeQueue().removeFirst();
+                protectedFrameProvider->imageFrameDecodeAtIndexHasFinished(request.index, request.subsamplingLevel, request.animatingState, request.options, WTFMove(nativeImage));
+            });
+        }
+
+        // Ensure destruction happens on creation thread.
+        callOnMainThread([protectedThis = WTFMove(protectedThis), protectedWorkQueue = WTFMove(protectedWorkQueue), protectedFrameProvider = WTFMove(protectedFrameProvider)] () mutable { });
```

Comment:
You're moving existing code here, so no need to do this now, but you can also achieve this by making these classes inherit from `ThreadSafeRefCounted<..., DestructionThread::Main>`.

Reply:
That is true.